pkg/auth/authmap_gc.go

```diff
@@ -0,0 +1,174 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package auth
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/cilium/cilium/pkg/identity"
+	ciliumv2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/node/addressing"
+)
+
+type authMapGarbageCollector struct {
+	authmap authMap
+	ipCache ipCache
+
+	discoveredCiliumNodeIDs    map[uint16]struct{}
+	discoveredCiliumIdentities map[identity.NumericIdentity]struct{}
+}
+
+func newAuthMapGC(authmap authMap, ipCache ipCache) *authMapGarbageCollector {
+	return &authMapGarbageCollector{
+		authmap: authmap,
+		ipCache: ipCache,
+		discoveredCiliumNodeIDs: map[uint16]struct{}{
+			0: {}, // Local node 0 is always available
+		},
+		discoveredCiliumIdentities: map[identity.NumericIdentity]struct{}{},
+	}
+}
+
+func (r *authMapGarbageCollector) HandleCiliumNodeEvent(_ context.Context, e resource.Event[*ciliumv2.CiliumNode]) error {
+	switch e.Kind {
+	case resource.Upsert:
+		if r.discoveredCiliumNodeIDs != nil {
+			log.Debug("auth: nodes discovered - getting node id")
+			remoteNodeIDs := r.remoteNodeIDs(e.Object)
+			for _, rID := range remoteNodeIDs {
+				r.discoveredCiliumNodeIDs[rID] = struct{}{}
+			}
+		}
+	case resource.Sync:
+		log.Debug("auth: nodes synced - cleaning up missing nodes")
+		if err := r.cleanupMissingNodes(); err != nil {
+			e.Done(err)
+			return fmt.Errorf("failed to cleanup missing nodes: %w", err)
+		}
+		r.discoveredCiliumNodeIDs = nil
+	case resource.Delete:
+		log.Debugf("auth: node deleted - cleaning up: %s", e.Key.Name)
+		if err := r.cleanupDeletedNode(e.Object); err != nil {
+			e.Done(err)
+			return fmt.Errorf("failed to cleanup deleted node: %w", err)
+		}
+	}
+	e.Done(nil)
+	return nil
+}
+
+func (r *authMapGarbageCollector) HandleCiliumIdentityEvent(_ context.Context, e resource.Event[*ciliumv2.CiliumIdentity]) error {
```

Comment:
`CiliumIdentity` is only valid when identity mode is CRD. This won't work with kvstore.

Reply:
@joamaki yes, this is mentioned in the PR description. will be tracked and refactored towards an abstraction which covers kvstore & crd based backend.

btw: as discussed offline - resource cells cilium nodes & cilium identity are now provided privately only within the auth cell.